applications/AdjointFluidApplication/custom_utilities/numerical_diffusion.h

```diff
@@ -0,0 +1,438 @@
+        }
+        else
+        {
+            KRATOS_THROW_ERROR(std::runtime_error,
+                "numerical diffusion method only supports 2D or 3D elements",
+                "")            
+        }        
+
+        pCurrentElement->SetValue(NUMERICAL_DIFFUSION, numerical_diffusion);
+
+        KRATOS_CATCH("");
+    }
+    ///@}
+
+private:
+    ///@name Member Variables
+    ///@{
+    numericalDiffusionMethod mNumericalDiffusionMethod = numericalDiffusionMethod::singularValuePressureCoupled;
+    double mBeta = 0.0;
+    ///@}
+    ///@name Private Operators
+    ///@{
+    template<unsigned int TDim>
+    void InitializeMatrix(MatrixType& rAdjointMatrix)
+    {
+        KRATOS_TRY;
+
+        constexpr unsigned int TNumNodes = TDim + 1;        
+        constexpr unsigned int TBlockSize = TDim + 1;    
+        constexpr unsigned int TFluidLocalSize = TBlockSize * TNumNodes;    
+        
+        if (rAdjointMatrix.size1() != TFluidLocalSize || rAdjointMatrix.size2() != TFluidLocalSize)
+            rAdjointMatrix.resize(TFluidLocalSize,TFluidLocalSize,false);
+
+        for (IndexType i=0; i < TFluidLocalSize; i++)
+            for (IndexType j=0; j < TFluidLocalSize; j++)
+                rAdjointMatrix(i,j) = 0.0;
+
+        KRATOS_CATCH("")
+    }
+
+    template<unsigned int TDim>
+    void CalculateVelocityGradientTensor(
+        Element::Pointer pElement,
+        const ProcessInfo& rCurrentProcessInfo,
+        MatrixType& rMatrix
+    )
+    {
+        constexpr unsigned int TNumNodes = TDim + 1;        
+
+        if (rMatrix.size1() != TDim || rMatrix.size2() != TDim)
+            rMatrix.resize(TDim,TDim,false);
+        
+        rMatrix.clear();
+
+        boost::numeric::ublas::bounded_matrix<double, TNumNodes, TDim> DN_DX;
+        array_1d< double, TNumNodes > N;
+        double Volume;
+        GeometryUtils::CalculateGeometryData(pElement->GetGeometry(),DN_DX,N,Volume);
+
+        std::vector<array_1d<double, TDim>> rNodalVelocityVectors;
+        for (unsigned int iNode = 0; iNode < TNumNodes; iNode++)
+            rNodalVelocityVectors.push_back(pElement->GetGeometry()[iNode].FastGetSolutionStepValue(VELOCITY));
+
+        for (unsigned int i = 0; i < TDim; i++)
+            for (unsigned int j = 0; j < TDim; j++)
+                for (unsigned int k = 0; k < TNumNodes; k++)
+                    rMatrix(i,j) += DN_DX(k,i)*rNodalVelocityVectors[k][j];
+
+    }
+
+    template<unsigned int TDim>
+    double CalculateNumericalDiffusionSVMethodPressureCoupled(Element::Pointer pCurrentElement, const ProcessInfo& rCurrentProcessInfo)
+    {
+        constexpr unsigned int TNumNodes = TDim + 1;        
+
+        boost::numeric::ublas::bounded_matrix<double, TNumNodes, TDim> DN_DX;
+        array_1d< double, TNumNodes > N;
+        double Volume;
+
+        GeometryType& rGeom = pCurrentElement->GetGeometry();
+
+        GeometryUtils::CalculateGeometryData(rGeom,DN_DX,N,Volume);
+
+        MatrixType GradVel;
```

Comment:
grad_vel. Probably it is GradVel in the vms adjoint element, but this is not the consistent code style of kratos.